### PR TITLE
changed PSE 01 to most frequent k elements

### PR DESCRIPTION
**Pull request by mikellewade**

### CODE REVIEW

File: 01/most-frequent-k-elements.md
Comment (yangashley) [inline 01/most-frequent-k-elements.md:1]: I'm not able to comment on lines of code not part of the commit so calling out my comments here. 

Line 64: says "Write unit tests for `pascals_triangle`" but should say "Write unit tests for `most_frequent_k_elements`"

Line 127-134: stray test called `test_pascals_out_of_range` should be removed

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        arr = [1,1,1,2,2,3]
+        k = 2
+
+        # Act
+        result = most_frequents_k_elements(arr, k)
+
+        # Assert
+        self.assertEqual(result, [1,2])
+
+    def test_pascals_big_number(self):
```
Comment (yangashley) [inline 01/implement-most-frequent-k-elements.checkpoint.md:156]: Should we rename the test to something like `test_arr_with_one_element` since students aren't working on Pascal's Triangle anymore?

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        k = 1 
+
+
+        # Act
+        result = most_frequent_k_elements(arr, k)
+
+        # Act
+        self.assertEqual(result, [1])
+
+    def test_pascals_negative_number_case(self):
```
Comment (yangashley) [inline 01/implement-most-frequent-k-elements.checkpoint.md:168]: Change name from `test_pascals` to `test_most_freq_negative_number_case` or something to that effect

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        arr = [-1,-1,-1,-2,-5,-5,3,3,3,3,4]
+        k = 3
+
+        # Act
+        result = most_frequent_k_elements(arr, k)
+
+        # Assert
+        self.assertEqual(result, [3,-1,-5])
+
+    def test_pascals_unordered_numbers_case(self):
```
Comment (yangashley) [inline 01/implement-most-frequent-k-elements.checkpoint.md:179]: Change name from `test_pascals` to `test_most_freq_unordered_number_case` or something to that effect

File: 01/most-frequent-k-elements.md
Comment (anselrognlie) [inline 01/most-frequent-k-elements.md:1]: As with Ashley's comments, I have a couple not on lines part of the commit.

I think the example clarifying questions (41,42) should be updated. How about:
1. Can the input array contain negative numbers?
2. What if there is a tie for the most frequent elements? E.g., what if `k` is 3, but there are two values tied for the 3rd most frequent element?
3. Will the input array always be in sorted order?
4. Must the output list be in a particular order?

Comment (anselrognlie) [reply 01/most-frequent-k-elements.md:1]: For `test_pascals_out_of_range`, rather than replacing, could we update it it to be a second (possibly edge) case? Something like:
```py
def test_unordered_numbers():
    # Arrange
    arr = [3, 1, 2, 3, 1, 3]
    k = 2 
    # Act
    result = most_frequent_k_elements(arr, k)

    # Assert
    assert result == [3, 1]
```

File: 01/implement-most-frequent-k-elements.instructor.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    for num in arr:
+        if hash_table.get(num, False):
+            hash_table[num] += 1
+            
+        else:
+            hash_table[num] = 1
+            result.append(num)
+            
+    i = 0
+    while i < len(result) - 1:
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.instructor.md:21]: Both the outer and inner loops (i and j) both run to completion every time. Prefer using euther _both_ `while` loops or _both_ `for` loops. In this case, I'd probably go with `for`. I think the `while` had been traditionally preferred in older CS Fun-dedicated lessons to avoid making algorithms look too pythonic, but I think we're fine here.

Comment (anselrognlie) [reply 01/implement-most-frequent-k-elements.instructor.md:21]: Alternatively, we could call sort(ed) to bring the time complexity down to O(n log n).

File: 01/implement-most-frequent-k-elements.instructor.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        for j in range(len(result) - 1):
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.instructor.md:22]: If we keep this written out, this looks like bubble sort to me? In which case, I think the inner loop only  needs to run up to `len(result) - i - 1`.

File: 01/implement-most-frequent-k-elements.instructor.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
+# Instructor: Implement Most Frequent K Elements
+
+An example of a working implementation:
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.instructor.md:3]: Maybe include time and space complexity? This looks like quadratic O(n^2) time and linear O(n) space to me.

File: 01/implement-most-frequent-k-elements.instructor.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+```python
+def top_k_frequent_elements(arr, k):
+    if len(arr) == 1: return [arr[0]]
+    
+    hash_table = {}
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.instructor.md:9]: Prefer a descriptive name like `counts` or `freqency_map` (what it's for) rather than `hash_map` (what it is).

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+        arr = [1,1,1,2,2,3]
+        k = 2
+
+        # Act
+        result = most_frequents_k_elements(arr, k)
+
+        # Assert
+        self.assertEqual(result, [1,2])
+
+    def test_pascals_big_number(self):
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.checkpoint.md:62]: Rename the other tests in here to refer to the most frequent k rather than pascal.

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+# Time complexity: O(N + (K - 1)^2) where N is length of the array and K is the number of unique integers.
+# Space complexity: O(N + K) where N is length of the array and K is the number of unique integers.
+def top_k_frequent_elements(arr, k):
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.checkpoint.md:107]: I'd bring over the suggestions I made on the instructor implementation here as well.

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+<summary>Click here to see the tests that will be run against your code</summary>
+
+```py
+    def test_most_k_frequent_elements_nominal(self):
+        # Arrange
+        arr = [1,1,1,2,2,3]
+        k = 2
+
+        # Act
+        result = most_frequents_k_elements(arr, k)
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.checkpoint.md:151]:
```suggestion
        result = most_frequent_k_elements(arr, k)
```

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+```
+Input: nums = [1], k = 1
+Output: [1]
+```
+
+### !end-question
+### !placeholder
+
+```python
+def most_frequent_k_elements(num_rows):
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.checkpoint.md:31]:
```suggestion
def most_frequent_k_elements(arr, k):
```

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+    def test_most_k_frequent_elements_nominal(self):
+        # Arrange
+        arr = [1,1,1,2,2,3]
+        k = 2
+
+        # Act
+        result = most_frequents_k_elements(arr, k)
+
+        # Assert
+        self.assertEqual(result, [1,2])
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.checkpoint.md:60]: Since no explicit result ordering was specified, can we do order-independent testing for the tests?

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<br>
+<details style="max-width: 700px; margin: auto;">
+<summary>Click here to see the tests that will be run against your code</summary>
+
+```py
+    def test_most_k_frequent_elements_nominal(self):
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.checkpoint.md:145]: Prefer pytest syntax throughout. No `self` param to the test, and instead of `self.assertCondition` calls, use `assert` and a regular comparison. We have to use `unittest` syntax for the tests that actually get run, but we should stick with pytest syntax for what the students see.

File: 01/implement-most-frequent-k-elements.instructor.md [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    result = []
+
+    for num in arr:
+        if hash_table.get(num, False):
```
[the remark this answers is not in the document]
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.instructor.md:13]: I thought about using 0 but I chose to go with False just in case 0 was an integer included in the given array. Also, I didn't know that `if key in dict` had a time complexity of O(1)! Thank you!

File: 01/implement-most-frequent-k-elements.instructor.md (hunk 4 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.instructor.md:21]: Okay thank you! I didn't know if they were allowed to use `sort`, I'll replace what I have with that!

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 12 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:60]: ✅

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 8 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:62]: ✅

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 13 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:145]: ✅

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 9 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:107]: ✅

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 11 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:31]: ✅

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 10 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:151]: ✅

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 1 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:156]: ✅

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 2 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:168]: ✅

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 3 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:179]: ✅

File: 01/implement-most-frequent-k-elements.instructor.md (hunk 7 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.instructor.md:9]: ✅

File: 01/implement-most-frequent-k-elements.instructor.md (hunk 5 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.instructor.md:22]: ✅

File: 01/most-frequent-k-elements.md
Comment (mikellewade) [reply 01/most-frequent-k-elements.md:1]: ✅

Comment (mikellewade) [reply 01/most-frequent-k-elements.md:1]: ✅

File: 01/most-frequent-k-elements.md [hunk 15]
```diff
@@ -155,15 +114,15 @@ class TestPython1(unittest.TestCase):
...
-
+    arr = [-1,-1,2,3,3,3]
+    k = 2 
     # Act
-    result = pascals_triangle(numRows)
+    result = most_frequent_k_elements(arr, k)
 
     # Assert
-    self.assertEqual(result, [[1], [1, 1]])
+    self.assertEqual(result, [3,-1])
```
[the remark this answers is not in the document]
Comment (mikellewade) [reply 01/most-frequent-k-elements.md:125]: ✅

File: 01/most-frequent-k-elements.md [hunk 16]
```diff
@@ -201,14 +160,10 @@ Write the logical steps here.
...
-    1. Initialize list for current row by setting the size to i + 1
-    2. Set the first and last values of the row to 1
-    3. Loop through the rest of the row
-        1. Calculate the current element in the row by adding the elements in the row above and to the left and above and to the right
-    4. Append the row list to the triangle list
-4. Return the triangle list
+2. Initialize hash that holds integer occurences
+3. Loop through unique integers
+    1. Sort them in descending order
+4. Return list
```
[the remark this answers is not in the document]
Comment (mikellewade) [reply 01/most-frequent-k-elements.md:166]: Could you give a little more detail as to what you mean by this and what you think should be here instead?

[the remark this answers is not in the document]
Comment (anselrognlie) [reply 01/most-frequent-k-elements.md]: In the classroom material, the sample approaches were listed as an explanation rather than a hint. Hints are available from the get go, while explanations aren't available until an answer is submitted. I wouldn't expect a "hint" to be the actual answer. If I were a student in this situation, I'd be disappointed if I only wanted a hint but was given the actual answer.

[the remark this answers is not in the document]
Comment (mikellewade) [reply 01/most-frequent-k-elements.md]: Oh okay okay, gotcha gotcha! I thought you meant like that section was suppose to be an explanation but I just need to not give them the answer!

File: 01/implement-most-frequent-k-elements.instructor.md (hunk 4 above)
Comment (anselrognlie) [reply 01/implement-most-frequent-k-elements.instructor.md]: If the problem isn't specifically to sort something (sorting here is only a portion of the approach) I wouldn't expect a student (or interview candidate) to implement a sort algorithm, especially when the sort provided in most languages is O(n log n), but a sort we'd implement on the spot is likely to be O(n^2).

Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.instructor.md]: Thank you!

File: 01/most-frequent-k-elements.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (anselrognlie) [reply 01/most-frequent-k-elements.md]: Oh yeah. Sorry. I just meant to use
```
##### !explanation 
##### !end-explanation 
```

rather than 
```
##### !hint
##### !end-hint 
```

File: 01/implement-most-frequent-k-elements.instructor.md [hunk 17]
```diff
@@ -0,0 +1,26 @@
+# Instructor: Implement Most Frequent K Elements
+
+An example of a working implementation:
+
+```python
+#Time Complexity: O(N log(N)) where N is the size of the array
+# Space Complexity: O(K) where K is the number of unique integers in a given array
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.instructor.md:7]: If picking a different variable, I'd go with M rather than K, to avoid confusion with the `k` param. But I'd probably stick with N for everything.

In the worst case, the number of unique numbers in the array is the same as the length of the array itself, so everything can be expressed in terms of the length of the array.

File: 01/most-frequent-k-elements.md [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+
+##### !end-question
+
+##### !placeholder
+
+Write the logical steps here.
+
+##### !end-placeholder
+
+##### !hint
```
Comment (anselrognlie) [inline 01/most-frequent-k-elements.md:166]: I like these hints, but can the original steps you had in the previous commit come back as the explanation?

File: 01/most-frequent-k-elements.md [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+    self.assertEqual(1,1)
+```
+
+##### !end-tests
+##### !explanation 
+
+Example tests:
+
+```python
+def test_negative_numbers(arr,k):
```
Comment (anselrognlie) [inline 01/most-frequent-k-elements.md:120]:
```suggestion
def test_negative_numbers():
```

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 12 above)
Comment (anselrognlie) [reply 01/implement-most-frequent-k-elements.checkpoint.md:60]: By order-independent, I imagined wrapping the result list in a set and comparing to an expected set, so that it doesn't matter whether the submitted implementation puts the results in order of most to least frequent, since order wasn't a stated requirement.

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 9 above)
Comment (anselrognlie) [reply 01/implement-most-frequent-k-elements.checkpoint.md:96]: Maybe leave off the complexity here, since this will probably be visible before a student writes up their own complexity analysis. Though if you wanted to include it as the explanations for the time and space complexity (clearly marked as the complexity of the sample implementation), I think that's fine.

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+   # looping through array and creating hash where key value pairs are unique integers and number of occurences
+    for num in arr:
+        if num in frequency_map:
+            frequency_map[num] += 1
+            
+        else:
+            frequency_map[num] = 1
+            uniques.append(num)
+            
+    # Sorted has time complexity of Nlog(N) and takes an interable, function to decide the order, and reverse to decide descending/ascending
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.checkpoint.md:110]:
```suggestion
    # Sorted has time complexity of Nlog(N) and takes an iterable, function to decide the order, and reverse to decide descending/ascending
```

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+# Time Complexity: O(N log(N)) where N is the size of the array
+# Space Complexity: O(K) where K is the number of unique integers in a given array
+def top_k_frequent_elements(arr, k):
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.checkpoint.md:96]:
```suggestion
def most_frequent_k_elements(arr, k):
```

File: 01/implement-most-frequent-k-elements.instructor.md [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+def top_k_frequent_elements(arr, k):
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.instructor.md:8]:
```suggestion
def most_frequent_k_elements(arr, k):
```

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+    if len(arr) == 1: return [arr[0]]
+    
+    frequency_map = {}
+    uniques = []
+   # looping through array and creating hash where key value pairs are unique integers and number of occurences
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.checkpoint.md:101]: indentation
```suggestion
    # looping through array and creating hash where key value pairs are unique integers and number of occurences
```

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 24]
```diff
@@ -0,0 +1,208 @@
...
+    for num in arr:
+        if num in frequency_map:
+            frequency_map[num] += 1
+            
+        else:
+            frequency_map[num] = 1
+            uniques.append(num)
+            
+    # Sorted has time complexity of Nlog(N) and takes an interable, function to decide the order, and reverse to decide descending/ascending
+   result = sorted(uniques, key=lambda num: frequency_map[num], reverse=True)
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.checkpoint.md:111]: indentation
```suggestion
    result = sorted(uniques, key=lambda num: frequency_map[num], reverse=True)
```

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 10 above)
Comment (anselrognlie) [reply 01/implement-most-frequent-k-elements.checkpoint.md:132]: Looks like this still needs to be applied.

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 25]
```diff
@@ -0,0 +1,208 @@
...
+from main import *
+
+class TestChallenge(unittest.TestCase):
+    def test_most_k_frequent_elements_nominal(self):
+        # Arrange
+        arr = [1,2,3,1,1,2,3]
+        k = 2
+
+        # Act
+        result = most_frequents_k_elements(arr, k)
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.checkpoint.md:57]:
```suggestion
        result = most_frequent_k_elements(arr, k)
```

File: 01/implement-most-frequent-k-elements.instructor.md [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+
+    for num in arr:
+        if num in frequency_map:
+            frequency_map[num] += 1
+            
+        else:
+            frequency_map[num] = 1
+            uniques.append(num)
+            
+   result = sorted(uniques, key=lambda num: frequency_map[num], reverse=True)
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.instructor.md:22]: indentation
```suggestion
    result = sorted(uniques, key=lambda num: frequency_map[num], reverse=True)
```

File: 01/most-frequent-k-elements.md [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+
+##### !placeholder
+
+Write the logical steps here.
+
+##### !end-placeholder
+
+##### !hint
+1. How could we use a hash/frequency map to keep track of how many times an integer shows up? 
+2. Is there a way we could use the frequency map to sort the integers based on their number of occurences?
```
Comment (anselrognlie) [inline 01/most-frequent-k-elements.md:168]:
```suggestion
2. Is there a way we could use the frequency map to sort the integers based on their number of occurrences? 
```

File: 01/implement-most-frequent-k-elements.instructor.md [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+
+    return result[:k]
+```
+
```
Comment (anselrognlie) [inline 01/implement-most-frequent-k-elements.instructor.md:26]: For instructor reference (for use in possible discussions) here's a (probablistically) O(N) time complexity approach. I say probablistically, since like quick sort, the partitioning here _could_ hit the degenerate case where it takes O(n^2) time. But this is rare. In practice, it will be closer to O(n) time. Note, this does not guarantee the _order_ of the top k, only that it will find them. If the result must be in decending order by frequency (not specified in the prompt), then the sort approach is probably preferable.
```py
# assuming we avoid the degenerate pivot case, this will tend towards O(n). the
# following argument assumes we get "good" pivot selection. In this case, the
# maximum number of outer iterations will be log n. during each iteration, we
# halve the number of elements being inspected. The first iteration, we inspect
# all n items. the next, n/2, then n/4, n/8, n/16, etc. The sum of the series of
# halves of n is n, plus the n of the first iteration is 2n, which is O(n).
def select_in_place(el_cnt_arr, k):
    start = 0
    end = len(el_cnt_arr)

    # if fewer than 2 items, there's no need to rearrange
    if end - start < 2:
        return

    # run until the k largest things are at the front of the array
    # (condition is checked internal to the infinite loop and returns when met)
    while True:
        pivot_pos = end - 1  # use final element as the pivot
        _, pivot_cnt = el_cnt_arr[pivot_pos]

        # track the end of the larger array (left), which is the eventual destination
        # of the pivot value
        dest_pos = start

        # compare each value with the pivot, starting at the beginning of the list
        for i in range(start, pivot_pos):
            _, cmp_cnt = el_cnt_arr[i]
            if cmp_cnt > pivot_cnt:
                # higher count than the pivot, so add to the larger array
                el_cnt_arr[dest_pos], el_cnt_arr[i] = el_cnt_arr[i], el_cnt_arr[dest_pos]

                # advance the pivot destination (also the end of the larger list)
                dest_pos += 1

        # move the pivot to its destination
        el_cnt_arr[dest_pos], el_cnt_arr[pivot_pos] = el_cnt_arr[pivot_pos], el_cnt_arr[dest_pos]

        # if we've found k larger things (minus 1 for 0-based index)
        if dest_pos == k-1:
            return
        elif dest_pos < k-1:
            # need more larger items to reach k, and they are to the right of the
            # current pivot destination, so shift start to the right
            start = dest_pos + 1
        else:
            # need fewer larger items to reach k, and they are in the region to
            # the left of the current pivot destination, so shift end to the left
            end = dest_pos
            
def select_top_k(items, k):
    arr_cp = list(items)  # copy (and make sure we even have a list) since we'll work in place
    select_in_place(arr_cp, k)  # make sure the k largest (el, cnt) are at the front
    result = [el for el, _ in arr_cp]  # unpack to just have the values
    return result[:k]  # take the first k

# Time Complexity: O(N) where N is the size of the array
# Space Complexity: O(M) where M is the number of unique integers in a given array (which is N in the worst case)
def most_frequent_k_elements(arr, k):
    if len(arr) == 1: return [arr[0]]
    
    frequency_map = {}
    uniques = []
    # looping through array and creating hash where key value pairs are unique integers and number of occurrences
    for num in arr:
        if num in frequency_map:
            frequency_map[num] += 1
            
        else:
            frequency_map[num] = 1
            uniques.append(num)
            
    # select_top_k tends towards O(n) assuming good pivot selection. though, like
    # quick sort, it could degenerate to O(n^2). unlike the use of sorting, there
    # is no guarantee made as to the order the top k will be returned in.
    result = select_top_k(frequency_map.items(), k)

    # use k to return the k most frequent integers
    return result[:k]
```

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 12 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:60]: OH. I understand now!

Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:60]: ✅

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 20 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:110]: ✅

File: 01/implement-most-frequent-k-elements.checkpoint.md (hunk 24 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:111]: ✅

File: 01/implement-most-frequent-k-elements.instructor.md (hunk 17 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.instructor.md:7]: ✅

File: 01/implement-most-frequent-k-elements.instructor.md (hunk 22 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.instructor.md:8]: ✅

File: 01/implement-most-frequent-k-elements.instructor.md (hunk 26 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.instructor.md:22]: ✅

File: 01/implement-most-frequent-k-elements.instructor.md (hunk 28 above)
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.instructor.md:26]: ✅

File: 01/implement-most-frequent-k-elements.checkpoint.md [hunk 29]
```diff
@@ -0,0 +1,206 @@
...
+
+
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+def most_frequent_k_elements(arr, k):
```
[the remark this answers is not in the document]
Comment (mikellewade) [reply 01/implement-most-frequent-k-elements.checkpoint.md:94]: I took it off because I was thinking it would spoil the answer for them, but this is for the example solutions not necessarily theirs. So I will put it back here, I feel like it would make more sense to see it here.

**Review by anselrognlie — APPROVED** [review]
👍 LGTM
